### PR TITLE
Resolve ambiguities in https://github.com/jonniedie/ComponentArrays.jl/pull/231

### PR DESCRIPTION
**Pull request by vpuri3**
To avoid conflict with

https://github.com/JuliaSparse/SparseArrays.jl/blob/main/src/sparsevector.jl#L1242-L1247

without this PR:
```julia
julia> vcat(ones(4), ComponentArray((;a = ones(2), b = ones(2))))
ERROR: MethodError: vcat(::Vector{Float64}, ::ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:4)}}}) is ambiguous.

Candidates:
  vcat(x::AbstractVector, y::ComponentVector)
    @ ComponentArrays ~/.julia/packages/ComponentArrays/7wTG6/src/array_interface.jl:37
  vcat(X::Union{Number, AbstractVecOrMat{<:Number}}...)
    @ SparseArrays ~/.julia/juliaup/julia-1.10.0-beta3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/SparseArrays/src/sparsevector.jl:1229

Possible fix, define
  vcat(::AbstractVector{<:Number}, ::ComponentVec
```

with this PR:
```julia
julia> vcat(ones(4), ComponentArray((;a = ones(2), b = ones(2))))
8-element Vector{Float64}:
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
```

### CODE REVIEW

**Comment by vpuri3** [comment]
ok tests are passing on < 1.10 now. Gonna test on 1.10 next

**Comment by vpuri3** [comment]
The `vcat` ambiguity in 1.10 is pretty crazy:

https://github.com/jonniedie/ComponentArrays.jl/actions/runs/6707386215/job/18225805848?pr=230#step:6:353

```julia
MethodError: vcat(::ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1, b = 2:3, c = ViewAxis(4:10, Axis(a = ViewAxis(1:3, Axis(a = 1, b = 2:3)), b = 4:7)))}}}, ::ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1, b = 2:3, c = ViewAxis(4:10, Axis(a = ViewAxis(1:3, Axis(a = 1, b = 2:3)), b = 4:7)))}}}, ::ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1, b = 2:3, c = ViewAxis(4:10, Axis(a = ViewAxis(1:3, Axis(a = 1, b = 2:3)), b = 4:7)))}}}) is ambiguous.
  
  Candidates:
    vcat(x::ComponentVector, args...)
      @ ComponentArrays ~/work/ComponentArrays.jl/ComponentArrays.jl/src/array_interface.jl:61
    vcat(x::ComponentVector, args::Union{Number, UniformScaling, AbstractVecOrMat}...)
      @ ComponentArrays ~/work/ComponentArrays.jl/ComponentArrays.jl/src/array_interface.jl:62
    vcat(x::ComponentVector, args::Vararg{AbstractVector{T}, N}) where {T, N}
      @ ComponentArrays ~/work/ComponentArrays.jl/ComponentArrays.jl/src/array_interface.jl:63
    vcat(X::Union{Number, AbstractVecOrMat{<:Number}}...)
      @ SparseArrays /opt/hostedtoolcache/julia/1.10.0-beta3/x64/share/julia/stdlib/v1.10/SparseArrays/src/sparsevector.jl:1229
    vcat(A::Union{Number, AbstractArray, UniformScaling}...)
      @ LinearAlgebra /opt/hostedtoolcache/julia/1.10.0-beta3/x64/share/julia/stdlib/v1.10/LinearAlgebra/src/uniformscaling.jl:428
    vcat(A::Union{AbstractArray, UniformScaling}...)
      @ LinearAlgebra /opt/hostedtoolcache/julia/1.10.0-beta3/x64/share/julia/stdlib/v1.10/LinearAlgebra/src/uniformscaling.jl:426
    vcat(A::Union{Number, AbstractArray}...)
      @ Base abstractarray.jl:1992
    vcat(A::AbstractArray...)
      @ Base abstractarray.jl:1991
    vcat(A::AbstractVecOrMat...)
      @ Base abstractarray.jl:1682
    vcat(V::AbstractVector...)
      @ Base abstractarray.jl:1612
    vcat(A::AbstractVecOrMat{T}...) where T
      @ Base abstractarray.jl:1683
    vcat(V::AbstractVector{T}...) where T
      @ Base abstractarray.jl:1613
  To resolve the ambiguity, try making one of the methods more specific, or adding a new method more specific than any of the existing applicable methods.
 ```

**Comment by vpuri3** [comment]
@jonniedie can you help with ^?

**Comment by vpuri3** [comment]
which packages are you talking about?

**Comment by vpuri3** [comment]
I made the methods more specific, and tests are passing locally

**Comment by vpuri3** [comment]
@jonniedie looks like this is good to go. Bumping version.

File: src/componentarray.jl [hunk 1]
```diff
@@ -115,21 +115,21 @@ ComponentMatrix{T}(x::ComponentMatrix) where {T} = T.(x)
 ComponentMatrix() = ComponentMatrix(Array{Any}(undef, 0, 0), (FlatAxis(), FlatAxis()))
 ComponentMatrix{T}() where {T} = ComponentMatrix(Array{T}(undef, 0, 0), (FlatAxis(), FlatAxis()))
 
-const CArray = ComponentArray
-const CVector = ComponentVector
-const CMatrix = ComponentMatrix
+const CArray{T} = ComponentArray{T} where{T}
+const CVector{T} = ComponentVector{T} where{T}
+const CMatrix{T} = ComponentMatrix{T} where{T}
```
Comment (jonniedie) [inline src/componentarray.jl:120]: Are these necessary? Type parameters should be already curried.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -18,6 +18,7 @@ jobs:
           - '1.6'
           - '1.8'
           - '1' # Leave this line unchanged. '1' will automatically expand to the latest stable 1.x release of Julia.
+          - '1.10.0-beta3'
```
Comment (jonniedie) [inline .github/workflows/ci.yml:21]:
```suggestion
          - 'beta'
```
I think this should run against the current beta release.

**Comment by vpuri3** [comment]
all comments have been resolved

**Comment by vpuri3** [comment]
@jonniedie can we merge?

File: src/componentarray.jl [hunk 3]
```diff
@@ -124,12 +124,12 @@ const AdjOrTransComponentArray{T, A} = Union{Adjoint{T, A}, Transpose{T, A}} whe
 const AdjOrTransComponentVector{T} = Union{Adjoint{T, A}, Transpose{T, A}} where A<:ComponentVector
 const AdjOrTransComponentMatrix{T} = Union{Adjoint{T, A}, Transpose{T, A}} where A<:ComponentMatrix
 
-const ComponentVecOrMat = Union{ComponentVector, ComponentMatrix}
-const AdjOrTransComponentVecOrMat = AdjOrTrans{T, <:ComponentVecOrMat} where T
-const AbstractComponentArray = Union{ComponentArray, AdjOrTransComponentArray}
-const AbstractComponentVecOrMat = Union{ComponentVecOrMat, AdjOrTransComponentVecOrMat}
-const AbstractComponentVector = Union{ComponentVector, AdjOrTransComponentVector}
-const AbstractComponentMatrix = Union{ComponentMatrix, AdjOrTransComponentMatrix}
+const ComponentVecOrMat{T} = Union{ComponentVector{T}, ComponentMatrix{T}} where{T}
```
Comment (jonniedie) [inline src/componentarray.jl:127]: Type parameters should already be curried. Is there a reason we need to explicitly show the first parameter?

**Comment by vpuri3** [comment]
Sorry for the delay, was busy with school work. @jonniedie addressed your latest comment

**Comment by vpuri3** [comment]
@jonniedie

**Comment by vpuri3** [comment]
@jonniedie

**Comment by vpuri3** [comment]
@jonniedie

**Comment by vpuri3** [comment]
@jonniedie ping

**Comment by jonniedie** [comment]
Thanks for doing this. One last ask, though: would you mind making a test case of the example you gave above so this will have coverage?

**Pull request merged**